batcher/sweep_batch.go

```diff
@@ -0,0 +1,1353 @@
+	confCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	confChan, errChan, err := b.chainNotifier.RegisterConfirmationsNtfn(
+		confCtx, b.BatchTxid, b.BatchPkScript, batchConfHeight,
+		b.currentHeight, lndclient.WithReOrgChan(reorgChan),
+	)
+	if err != nil {
+		b.log.Errorf("unable to register for confirmation ntfn: %v", err)
+		b.writeToErrChan(ctx, err)
+		return
+	}
+
+	for {
+		select {
+		case conf := <-confChan:
+			select {
+			case b.confChan <- conf:
+
+			case <-ctx.Done():
+			}
+			return
+
+		case err := <-errChan:
+			b.writeToErrChan(ctx, err)
+			return
+
+		case <-reorgChan:
+			// A re-org has been detected. We set the batch's state
+			// back to open since our batch transaction is no longer
+			// present in any block. We can accept more sweeps and
+			// try to publish new transactions, at this point we
+			// need to monitor again for a new spend.
+			b.sweepsLock.Lock()
+			b.State = Open
+			b.sweepsLock.Unlock()
+
+			b.log.Warnf("reorg detected")
+			return
+
+		case value := <-b.confRefreshChan:
+			// If the nonce we received from the channel is equal
+			// to this routine's nonce, this means that we are the
+			// first routine to arrive here. So we only exit if the
+			// nonce we read is different.
+			if value != nonce {
+				return
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// handleSpend handles a spend notification.
+func (b *Batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+	// As soon as a spend arrives we should no longer accept sweeps. Also,
+	// the closed state will cause the batch to start monitoring the
+	// confirmation of the batch transaction.
+	b.sweepsLock.Lock()
+
+	b.State = Closed
+	var (
+		txHash       = spendTx.TxHash()
+		purgeList    = make([]SweepRequest, 0, len(b.sweeps))
+		spendCounter int
+	)
+	b.BatchTxid = &txHash
+	b.BatchPkScript = spendTx.TxOut[0].PkScript
+
+	// As a previous version of the batch transaction may get confirmed,
+	// which does not contain the latest sweeps, we need to detect the
+	// sweeps that did not make it to the confirmed transaction and feed
+	// them back to the batcher. This will ensure that the sweeps will enter
+	// a new batch instead of remaining dangling.
+	for _, sweep := range b.sweeps {
+		found := false
+
+		for _, txIn := range spendTx.TxIn {
+			if txIn.PreviousOutPoint == sweep.Outpoint {
+				found = true
+				spendCounter++
+			}
+		}
+
+		// If the sweep's outpoint was not found in the transaction's
+		// inputs this means it was left out. So we delete it from this
+		// batch and feed it back to the batcher.
+		if !found {
+			newSweep := sweep
+			delete(b.sweeps, sweep.SwapHash)
+			purgeList = append(purgeList, SweepRequest{
+				SwapHash: newSweep.SwapHash,
+				Outpoint: newSweep.Outpoint,
+				Value:    newSweep.Value,
+			})
+		}
+	}
+
+	b.sweepsLock.Unlock()
+
+	// If we have sweeps to purge, we send them back to the batcher.
+	for _, sweep := range purgeList {
+		b.returnChan <- sweep
```

Comment:
I believe this could cause a deadlock if the batcher quits right when we're trying to send to the `returnChan`. Would be nicer to express the "return" as a function of the `Batcher` that also ensures that once the `Batcher` is starting to quit we can't write to the `returnChan`.

Reply:
oh right, yeah I need to listen for the `ctx.Done` here too

> express the "return" as a function of the Batcher that also ensures that once the Batcher is starting to quit we can't write to the returnChan

that's doable more directly if we just listen to the context. If batcher quites the context exits and so do the batches